### DISCUSSION TITLE
problems with math output

### DISCUSSION THREAD

**Discussion by philschrad** [open-webui/open-webui#3127] [General]
**Bug Summary:**
math output not rendering correctly 

**Steps to Reproduce:**
ask gpt4, for example: 

```
repeat after me: The transformation matrix for a rotation by an angle \(\theta\) counterclockwise is given by:

\[
\begin{pmatrix}
\cos(\theta) & -\sin(\theta) \\
\sin(\theta) & \cos(\theta)
\end{pmatrix}
\]

```
(this is also part of the response if you just ask for the transformation matrix)

**Expected Behavior:**
The math should display correctly, i.e.  the greek letter theta and a two by two matrix.

**Actual Behavior:**
When displayed, the backslashes on the math delimiters `\[ \]` and `\( \)` just disappear, and then the latex does not render. By editing the response I have also noticed the following behaviour: 
- double dollar signs work as inline math delimiters but single ones do not unless they are given whitespace, which for the single ones I think is not typical usage
-` \[` and `\]` work as inline delimiters only if given whitespace. 
-` \\[ ... \\]` and `\\( ... \\)` work as inline delimiters (without whitespace) but are not typical outputs from gpt?
- none of the delimiters work over multiple lines, like the code for the matrix above, I think this is because in [here](https://github.com/open-webui/open-webui/blob/main/src/lib/components/chat/Messages/ResponseMessage.svelte) all the delimiters are set with `display: false`. It would be good to have at least one way of rendering displayed equations. 
- even when the above code for the matrix is changed to be inline with working delimiters, it will display a row vector instead of a 2 by 2 matrix because the double backslash somehow gets ignored (unless I change it to a triple backlash). 

## Environment

- **Open WebUI Version:** not sure, whichever one installs when you follow the docker installation for OpenAI API usage only from [here](https://docs.openwebui.com/getting-started/)
- **Operating System:** [macOS Sonoma]
- **Browser (if applicable):** [Safari]

**Comment by AlanRvk** [comment]
It's a problem with multiline latex, I don't why but as a temporary solution I ask in the system prompts to only make single line latex, but the matrices and such wont work

**Comment by Algorithm5838** [comment]
Also, it can render the first 2 lines, but not the last 2:
```
- \[ 5 \text{ tables} \times 4 \text{ chairs per table} = 20 \text{ chairs} \]
- \[ 22 \text{ guests} - 20 \text{ chairs} = 2 \text{ guests} \]
- Total seating capacity = 20 (which is \( 5 \text{ tables} \times 4 \text{ chairs per table} \))
- \[ 22 \% 20 \]
```

**Comment by bohrium2b** [comment]
It is also broken with inline/single-line latex (so `$1 + 3$ or $3 + 1$` would be rendered as \$1 + 3 $or$ 3 + 1\$)

**Comment by chrisoutwright** [comment]
I had to add a heavy instruction and several attempts to get some useful display:
![image](https://github.com/open-webui/open-webui/assets/27736055/53e25f9c-34c9-4154-b2c0-c2096d1dc2ee)

but it needed several attempt, and most times it was plain worse when the instruction was a bit different:
![image](https://github.com/open-webui/open-webui/assets/27736055/e0caed37-1cb3-4572-8949-5714871dc43e)

**Reply by bohrium2b** [reply]
Thank you for helping to debug this! I feel like this is less an issue with the models themselves (and it would be difficult to change this), and open-webui should perhaps consider implementing #2615 , or at least changing the delimiters to match the default behaviour for the models.

**Reply by chrisoutwright** [reply]
So this should normally not have been rendered like this?:
I thought that it would work WITH the whitespace, but it still is not being displayed correctly

```
In the context of Dreamer-v3, \( E_{q_\phi} \) refers to the expectation operator taken over the distribution \( q_\phi \). This distribution usually represents the learned probability distribution that the algorithm uses to model various aspects of its environment and internal states.

When you see \( E_{q_\phi}(...) \) in an equation like \( L(\phi) = E_{q_\phi}[...]\), it means we are calculating the expected value or average result of the expression inside the brackets, considering all possible outcomes according to the probability distribution \( q_\phi \).

In the case of Dreamer-v3:

1. **\( q_\phi(z_t | h_t, x_t) \)**: This is typically the posterior distribution over latent variables (or representations), given by the current hidden state and input observations.

2. **\( q_\phi(zt | ht, xt) \)** appears within the dynamics loss \( L_{\text{dyn}}(\phi) \). Here it represents the distribution of predicted next latent states based on the current hidden state and observation.

3. The expectation is taken over this distribution when calculating losses such as \( L_{\text{pred}}(\phi), L_{\text{dyn}}(\phi), \) and \( L_{\text{rep}}(\phi) \).

Overall, \( E_{q_\phi}(...) \) is used to compute the average loss that the world model experiences over all possible states and actions it might encounter during its operation. This formulation ensures that the algorithm's learning is not biased towards any particular scenario but instead considers a wide range of potential outcomes based on the distribution it has learned, leading to more robust and generalizable performance across different environments or tasks.
```
![image](https://github.com/user-attachments/assets/9562da12-b730-49e6-9c6a-f6d36846e750)

**Reply by Yanyutin753** [reply]
https://github.com/open-webui/open-webui/pull/3930 Look at this. I tried to solve this problem
